README.md
[![Build Status](https://travis-ci.org/nrstott/bogart.png?branch=master)](https://travis-ci.org/nrstott/bogart)
# Bogart

A blazing fast rapid application development web framework using JSGI for [node](http://nodejs.org/).

## Getting Started

Bogart can be installed via [npm](https://github.com/isaacs/npm).

    npm install bogart

Alternatively, clone the git repository.

    git clone git://github.com/nrstott/bogart.git

## Hello World in Bogart

Make a directory: `mkdir hello-world`

Create the following file:

### app.js

```javascript
var bogart = require('bogart');

var router = bogart.router();
router.get('/', function(req) { 
  return "hello world"; 
});

router.get('/:name', function(req) {
  return 'hello '+req.params.name;
});

var app = bogart.app();
app.use(bogart.batteries); // A batteries included JSGI stack including streaming request body parsing, session, flash, and much more.
app.use(router); // Our router

app.start();
```

Start your app: `node app.js`

Visit it in a web browser at [http://localhost:8080](http://localhost:8080).
Visit the route that says hello to you by name at [http://localhost:8080/bob](http://localhost:8080/bob)

## Changing the port

If you can't run on 8080, change the `app.start` call e.g. `app.start(9090, '127.0.0.1')`

## Routing

Routing in Bogart is simple and intuitive.  A route is an HTTP method paired with an
URL matching pattern and a handler function.

```javascript
var router = bogart.router();
router.get('/', function(req) {
  return bogart.html('Hello World');
});
```

Routes are tested for matches in the order in which they are defined.

### Route Patterns

Route patterns are matched against URLs.  They may include named parameters that will
be accessible via the `params` object of the `req` object passed to the route handler.

```javascript
var router = bogart.router();
router.get('/hello/:name', function(req) {
  var greeting = 'Hello '+req.params.name;
  return bogart.html(greeting);
});
```

Route patterns support wildcards. Wildcards will match anything whereas regular named parameters
will not match beyond a path separator ("/").

```javascript
var router = bogart.router();
router.get('/hello/*', function(req, name) {
    return bogart.html('Hello '+req.params.splat[0]);
});
```

### Regex Routes

When a route pattern is not powerful enough, regular expressions may be used to specify which
URLs are to be matched.

```javascript
var router = bogart.router();
router.get(/\/posts?/, function(req) {
  // Matches 'post' or 'posts'
  return bogart.html('Regex Route');
});
```

Parameters are via regular expression groups in regular expression routes.  The parameter values
are put in an `Array` in `req.params.splat` of the `req` object passed to the route handler.

```javascript
var router = bogart.router();
router.get(/hello-(.*)/, function(req) {
  var name = req.params.splat[0];
  return bogart.html('Hello '+name);
});
```

### Custom HTTP Verbs

To handle HTTP Verbs other than `get`, `post`, `put`, and `delete`, the `route` method of 
`Router` can be invoked.

```javascript
var router = bogart.router();
router.route('options', function (req) {
  // Logic for handling the request goes here.
});
```

## Bogart Application

`bogart.app` makes it easy to setup a middleware chain and start coding. Combined with
`bogart.batteries` (See Below), you can setup a full-stack JSGI application in two lines of code.

```javascript
var app = bogart.app();
app.use(bogart.batteries);
```

After adding `bogart.batteries`, you will normally want to add a Router. This is also done
with `app.use`. To start the application, use the `start` method.

```javascript
var app = bogart.app();
app.use(bogart.batteries);

var router = bogart.router();
// NOTE: Here you would normally add some routes.

app.use(router);
app.start();
```

## Response Helpers

Bogart includes helpers for creating JSGI Responses. The helpers, by convention, take a final
parameter of an options object that allows the caller to override defaults. The options object
is merged with the default JSGI Response values of the helper before the JSGI Response is returned.

### Respond with JSON

Helper to create an HTTP 200 Success response with a Content-Type of application/json.

Sample Route:

```javascript
var router = bogart.router();
router.get('/', function(req) {
  return bogart.json({ framework: 'Bogart' });
});
```

This route yields the following JSGI Response:

```
{
  status: 200,
  headers: { 'Content-Type': 'application/json' },
  body: [ '{ "framework": "Bogart" }' ]
}
```

### Redirect

Helper to create an HTTP 302 Temporary Redirect response.

Sample Route:

```javascript
var router = bogart.router();
router.get('/', function(req) {
  return bogart.redirect('/some/other/url');
});
```

This route yields the following JSGI Response:

```
{
  status: 302,
  headers: { 'Location': '/some/other/url' },
  body: []
}
```

### Error

Helper to create an HTTP 500 Internal Server Error response.

Sample Route:

```javascript
var router = bogart.router();
router.get('/', function(req) {
  return bogart.error('<html>...</html>');
});
```

This route yields the following JSGI Response:

```
{
  status: 500,
  headers: { 'Content-Type': 'text/html' },
  body: [ '<html>...</html>' ]
}
```

### Not Modified

Helper to create an HTTP 304 Not Modified response.

Sample Route:

```javascript
var router = bogart.router();
router.get('/', function(req) {
  return bogart.notModified();
});
```

This route yields the following JSGI Response:

```
{
  status: 304,
  headers: {},
  body: []
}
```

### File (Streaming!)

Helper to create an HTTP 200 Success response with a body streamed from the contents of a file. The
Content-Type of the response is determined by the mime type of the file.

Sample Route:

```javascript
var path = require('path');

var router = bogart.router();
router.get('/download/*', function(req) {
  var filePath = path.join(__dirname, 'public', req.params.splat[0]);

  return bogart.file(filePath);
});
```

This route yields the following JSGI Response:

```
{
  status: 200,
  headers: { 'Content-Type': '<mimetype of the file>' },
  body: fileStream // <-- A file stream
}
```

### Proxy (Streaming!)

Helper to create a response that proxies the response from a URL.

Sample Route:

```javascript
var router = bogart.router();
router.get('/google', function(req) {
  return bogart.proxy('http://www.google.com');
});
```

This route yields a JSGI Response that matches the response from the proxied URL.

## Imperative Response Builder

Bogart includes the `ResponseBuilder` helper to provide an imperative interface. While not
recommended as a goto style of programming in Bogart, there are times when buiding a
response imperatively makes for cleaner, better code. This is true especially
when working with callback based functions that cannot be wrapped by `bogart.promisify`.

```javascript
var router = bogart.router();
router.get('/', function(req) {

  // Get a ResponseBuilder
  var res = bogart.res();

  doSomethingAsync(function(err, messageStr) {
    res.setHeader('Content-Type', 'text/plain');

    if (err) {
      res.status(500);
      res.send('Error');
      return res.end(); // We use return to break out of the function, do not want to continue executing after res.end()
    }

    res.status(200);
    res.send(messageStr);
    res.end(); // End the Response.  This is analagous to resolving a promise for a JSGI Response.
  });

  return res;
});
```

## Using Session

The session middleware can be included individually with `app.use(bogart.middleware.session)` or
by using batteries `app.use(bogart.batteries)` which includes a default stack of JSGI middleware.

A `session` function will be available on the request object passed to your route handlers. This
function follows the jQuery style of arity determining if it is getting or setting a key/value pair.
A call to session with one argument is a get to the value of the key referenced by the argument.

```javascript
req.session('name'); // => value associated with 'name'
```

A call to session with two arguments is a set.

```javascript
req.session('name', 'Nathan'); // sets the value of 'name' to 'Nathan'
```

### Contrived Example

A set of two routes that use session:

```javascript
router.get('/:name', function(req) {
  req.session('name', req.params.name);
  return bogart.redirect('/');
});

router.get('/', function(req) {
  return bogart.html('Hello ' +req.session('name'));
});
```

Visiting '/:name' ('/Nathan', '/Bob', etc...) will set a session key that will be displayed by the root
route '/', after the redirect.

## Running the Examples

In the 'examples' directory of the cloned source, there are several examples of bogart applications.

### Hello World

The hello world example demonstrates a basic bogart applications.  The application has a route, '/:name', that takes a name
as a parameter in the request URL.  The application responds with 'hello <name>'.  It also has a route, '/', that responds with `hello world`.

    > cd examples
    > node hello-world.js

Visit the application in your web browser at [http://localhost:8080/Jim](http://localhost:8080/Jim)

### Mustache Template with Layout

This example demonstrates usage of the mustache templating engine and a mustache layout.  A layout is a page designed
to hold other templates to avoid duplication of content.

    > cd examples/mustache-layout
    > node app.js

Visit the application in a web browser at [http://localhost:8080/](http://localhost:8080)

### Jade

If you would like to use Jade instead of Mustache, please `npm install bogart-jade`. Then in
your application add `require('bogart-jade')` and the Jade ViewEngine will be available via
`bogart.viewEngine('jade')`. See [the github repository](https://github.com/nrstott/bogart-jade) for more information.

## Middleware
Bogart comes with a variety of JSGI middleware appliances. The `batteries` appliance includes a recommended JSGI application chain for
normal usage.

#### Batteries
Frameworks are better when they come batteries included. Bogart ships with the batteries, it's just up to you to put them in if you
want to use them.

The batteries middleware creates the following JSGI chain where `nextApp` is the JSGI appliance passed as the first parameter
to `batteries`:

    error -> validateResponse -> directory -> parted -> methodOverride
          -> session -> flash -> bodyAdapter -> stringReturnAdapter -> nextApp

#### Serve Static Files
The static example demonstrates using bogarts *Directory* middleware to serve an image.

    > cd examples/static-server
    > node app.js

Visit the application in a web browser at [http://localhost:8080/](http://localhost:8080).
You should see the image.

#### Error
Translates rejected promises to a JSGI error response.

#### Flash
Provides Ruby-esque 'flash' variables, good only for the next request cycle. Flash session tracking and flash data is stored via encrypted cookies by default. These providers can be overriden with the Flash middleware options. An example flash data provider implementation with a Redis backend is available here: [https://github.com/jdc0589/bogart-flash-redis](https://github.com/jdc0589/bogart-flash-redis)

#### MethodOverride
Checks for the _method parameter or X-HTTP-METHOD-OVERRIDE header to override the HTTP method.

#### ParseJson
Parses the body of any request with an "application/json" content-type. The value of request body is reassigned with the parsed value.
Use the Parted middleware instead unless you have a specific reason not to as it contains a superset of this functionality and also performs streaming parsing.

#### ParseForm
Parses the body of any request with an "application/x-www-form-urlencoded" content-type. The value of request body is reassigned with the parsed value.
Use the Parted middleware instead unless you have a specific reason not to as it contains a superset of this functionality and also performs streaming parsing.

#### Parted
A JSGI wrapper around the excellent [Parted](https://github.com/chjj/parted) middleware for Connect.

CHANGES.md
## v0.5.9

* Ignores `examples` directory in npm releases.

## v0.5.8

* Uses default `directory` middleware configuration if `config` is provided but no 
  no value for `directory` is present in call to `bogart.batteries`.

## v0.5.7

* Updates Q dependency to version 0.9.6. This version of Q has many performance enhancements.

## v0.5.6

* Updates static server example to use `bogart.app`.
* Switches Travis configuration to only run tests in node v0.10.x.
* Updates stream specs for node v0.10.x style streams.
* Fixes a bug with `bogart.pump` resolving its promise at wrong time.

## v0.5.5

* Falls back to `process.nextTick` when setImmediate is unavailable in `ForEachStream`.
* Resolves `ResponseBuilder` promise for the response on first call to `send` to support streaming.
* Add node v0.10.x to Travis configuration.

## v0.5.4

* Changes ForEachStream to use `setImmediate` instead of `process.nextTick`.

## v0.5.3

* Fixes bug preventing passing of options to `bogart.batteries`.

## v0.5.2

* Updates examples to use `bogart.app`.
* Normalizes the return value of `bogart.router` to match other middleware.
* Changes `bogart.error` middleware to match middleware conventions.
* Updates `bogart.session` example to use a custom configuration key.
* Allows configuration of session's `SessionIdProvider` and `SessionDataProvider`
  encryption key from the options parameter to `bogart.session`.
* Removes `bogart.build`.
* Removes `binary` middleware.
* Removes `cascade` middleware.

## v0.5.1

* `SessionDataProvider.loadSession` now returns the session.

## v0.5.0

* Changes order of parameters to `bogart.parted` middleware to match middleware conventions.
* Creates a helper module called `fsp` where node.js `fs` methods are adapted to return promises.
* General housekeeping on the middleware code base to make it more maintainable.
* Changes `bogart.middelware` to be a function that makes creating JSGI middleware easier. 
  All bogart middleware is still attached to the `bogart.middleware` function as properties.
* Removes the deprecated support for passing a configuration function to `bogart.router`.

## v0.3.38

* Changed `bogart.response()` to `bogart.res()` to better fit design philosophy and Node conventions.

## v0.3.37

* Changed signature of view engine renderer function so that the caching mechanism is readily available for view engine implementations.
* ResponseBuilder helper, `bogart.response()`, can now be piped to successfully.

## v0.3.36

* Publishing v0.3.35 from windows seemed to cause issues when installing on unix. Therefore; I am republishing with only a version bump from linux.

## v0.3.35

* Back out change to package.json to add new CLI.

## v0.3.34

* Removed dependency on Q library.

## v0.3.33

* Changed the unit tests in the project from Expresso to Tap.

## v0.3.32

* Did not update package.json in my tag of 0.3.31, went to 0.3.32 for consistency.

## v0.3.31

* Fixed a bug where Parted middleware was not bubbling rejections.

## v0.3.30

* Fixed a bug causing before callbacks to cause errors.

## v0.3.29

* Missed a whenCall, same issue from v0.3.28.

## v0.3.28

* Added dependency on request 2.2.9.
* Removed dependency on Deflate as Node.JS 0.6.x includes zlib.
* Added gzip middleware to `bogart.batteries`
* Added reject callbacks for all cases where whenCall is invoked as it tries to invoke the rejectCallback even if one is not provided.

## v0.3.27

* `bogart.middleware.session` assumed that `req.env` would be unique per request; however, it is not. Corrected issues caused by this.

## v0.3.26

* `bogart.middleware.bodyAdapter` now adapts responses that are of type Buffer or Stream to JSGI responses.
* Fixed a bug in `bogart.middleware.bodyAdatper` where Stream returns were not being handled properly.
* Added `bogart.config`. The default environment is 'development' and may be overridden with the BOGART_ENV environment variable.
* Expose `DefaultIdProvider` and `DefaultDataProvider` as properties of `bogart.middleware.session`.

## v0.3.25

* No longer reject a return that does not include all properties of a valid JSGI response.
  This change facilitates Bogart as a middleware platform.
* Added `bogart.middleware.batteries`, a batteries included JSGI stack for rapid application development.
* The deprecated `bogart.app` has been reclaimed for the purpose of creating application stacks more easily than chaining
  JSGI middleware manually or using `bogart.build`.
* `bogart.build` is deprecated.
* Added `bogart.q` which exposes the promise implementation used by Bogart.
* Added `bogart.promisify` which adapts node-style asynchronous functions to promises.
* Added `bogart.proxy`, a helper to create a JSGI response that proxies a URL.
* Added `viewEngine.share`, a helper for serializing JavaScript to views.  See the new example in 'examples/share-javascript'.

## v0.3.24

* Bug fixes in Session middleware.

## v0.3.23

* Added support for string-based paths with * for splat like /foo/*
* Before callbacks may now return promises that must be resolved before the route handler is executed.
* After callbacks may now return promises that must be resolved before the response from the route handler is returned.

## v0.3.22

* `ViewEngine` is now an `EventEmitter`.
* the built-in `Mustache` view engine now emits `beforeRender` and `afterRender` events.
* Fixed bugs in `Flash` middleware.

## v0.3.21

* `after` had been left off of the public API of router, added it.

## v0.3.20

* Simplified and corrected code for `pipe` method on the request object for the `Parted` middleware.
* Added `Session` middleware.

## v0.3.19

* Updated Parted dependency to 0.8.0.

## v0.3.18

* Mustache partials now work properly when using layouts.

## v0.3.17

* Routes now match in order added instead of longest-first.
* Added `Flash` middleware to emulate the flash method of Rails.

## v0.3.16

* Added `Error` middleware to translate rejected promises and thrown errors into an error response.

lib/bogart.js
var
  path         = require('path'),
  Q            = require('./q'),
  util         = require('./util'),
  Router       = require('./router').Router,
  middleware   = require('./middleware'),
  when         = Q.when,
  merge        = require('./util').merge,
  fs           = require('fs'),
  view         = require('./view'),
  inherits     = require('util').inherits,
  EventEmitter = require('events').EventEmitter;

exports.version = [0,5,9];

exports.util = require('./util');

exports.q = Q;

/**
 * Wraps a Node.JS style asynchronous function `function(err, result) {}` 
 * to return a `Promise`.
 *
 * @param {Function} nodeAsyncFn  A node style async function expecting a callback as its last parameter.
 * @param {Object}   context      Optional, if provided nodeAsyncFn is run with `this` being `context`.
 *
 * @returns {Function} A function that returns a promise.
 */
exports.promisify = Q.promisify;

exports.middleware = middleware;
exports.batteries  = middleware.batteries;

/**
 * Creates a request object given a router and a jsgi request.
 * This function is primarily intended to be used internally by bogart; however, it could be
 * used by a third party library to compose bogart routers with its own handling mechanisms.
 *
 * @type Request
 */
exports.request = require('./request');

/**
 * Creates a @see ViewEngine
 *
 * Example:
 *     bogart.viewEngine("mustache", require("path").join(__dirname, "/templates"))
 *
 * @param {String} engineName  the name of the engine, available: ["mustache", "haml"]
 * @param {String} viewsPath   Path where the views are located.  Defaults to /views
 * @member bogart
 */
exports.viewEngine = view.viewEngine;

/**
 * A JSGI Application
 *
 */
exports.app = function() {
  return new App();
};

function App() {
  EventEmitter.call(this);

  this.started = false;
  this.middleware = [];
}

inherits(App, EventEmitter);

App.prototype.use = function(middleware /*, parameters */) {
  if (this.started) {
    throw new Error('Application has already been started. Bogart application may only be configured before it is started.');
  }

  var args = Array.prototype.slice.call(arguments);
  middleware = args.shift();

  this.emit('beforeAddMiddleware', this, args);

  this.middleware.push({ middleware: middleware, args: args });

  this.emit('afterAddMiddleware', this, middleware);
};

App.prototype.start = function(port, host, jsgiOpts) {
  var app
    , first
    , args = Array.prototype.slice.call(arguments);

  jsgiOpts = args.shift();
  if (typeof jsgiOpts === 'object') {
    port = jsgiOpts.port || 8080;
    host = jsgiOpts.host || '127.0.0.1';
  } else {
    port = jsgiOpts;
    host = args.shift();
    jsgiOpts = args.shift() || {};
  }

  if (host === undefined) {
    host = port;
  }

  this.emit('beforeStart', this);

  this.middleware = this.middleware.reverse();
  first = this.middleware.shift();
  app = first.middleware.apply(first.middleware, first.args || []);

  if (this.middleware.length > 0) {
    this.middleware.forEach(function(descriptor) {
      app = descriptor.middleware.apply(descriptor.middleware, descriptor.args.concat([ app ]));
    });
  }

  var server = exports.start(app, require('./util').merge({
    port: port || 8080,
    host: host || '127.0.0.1'
  }, jsgiOpts));
  this.started = true;

  this.emit('afterStart', this, server);

  return server;
};

/**
 * Configuration manager. The function is variadic. The last argument must be a callback. The callback
 * will be executed if the BOGART_ENV environment variable matches one of the environments provided. If
 * no environments are provided, the function was called with one argument, then the environments default to
 * 'all'. The special string 'all' may also be passed to explicitly run the callback in all environments.
 *
 * Examples:
 *
 *     var app = bogart.app();
 *
 *     bogart.config(function() {
 *       // Executed in all environments
 *       app.use(bogart.batteries);
 *     });
 *
 *     bogart.config('development', function() {
 *       // Executed only when BOGART_ENV is set to 'development'
 *       app.use(requestLogger);
 *     });
 *
 *     bogart.config('staging', 'production', function() {
 *       // Executed in staging or production
 *       app.use(redisSession);
 *     });
 *
 * @returns undefined
 */
exports.config = function(/* environment1, environment2, ..., environmentN, callback */) {
  var args = Array.prototype.slice.call(arguments)
    , callback = args.pop()
    , environments = 'all';
  
  if (args.length > 0) {
    environments = args;
  }

  if (environments === 'all' || (environments.indexOf(process.env.BOGART_ENV || 'development') !== -1)) {
    callback();
  }
};

/**
 * Creates a bogart router.  A router is responsible for routing requests to appropriate handlers.
 *
 * Example:
 *     bogart.router(function(get, post, put, del) {
 *       get('/', function() { return bogart.html('Hello World'); });
 *     });
 *
 * Alternative:
 *     var router = bogart.router();
 *     router.get('/', function() { return bogart.html('Hello World'); });
 *
 * @param {Function} config       DSL configuration of routes.
 * @param {Function} notFoundApp  JSGI application to execute when no route from config matches the request.
 */
exports.router = Router;

/**
 * Starts a server
 *
 * @param {Function} jsgiApp   JSGI application to run
 * @param {Object} options     Options hash.  Supports 'port' property which allows specification of port for server.
 *                             Port defaults to 8080.  More options are planned for the future.
 */
exports.start = function(jsgiApp, options) {
  return require("jsgi").start(jsgiApp, options);
};

function jsgiResponse(body, opts, defaultHeaders) {
  defaultHeaders = defaultHeaders || {};

  opts = opts || {};
  opts.status = opts.status || 200;
  opts.headers = opts.headers || {};

  if (!Array.isArray(body) && !(typeof body.forEach === 'function')) {
    body = [ body ];
  }

  return {
    status: opts.status,
    body: body,
    headers: merge(defaultHeaders, opts.headers)
  };
}

/**
 * Text response.  Bogart helper method to create a JSGI response.
 * Returns a default JSGI response with body containing the specified text, a status of 200,
 * and headers.  The headers included are "content-type" of "text" and "content-length" set
 * appropriately based upon the length of 'txt' parameter.
 *
 * @param {String} txt  Text for the body of the response.
 */
exports.text = function(txt, opts) {
  txt = txt || '';

  return jsgiResponse(txt, opts, {
    'content-type': 'text',
    'content-length': Buffer.byteLength(txt, 'utf-8')
  });
};

/**
 * HTML response.  Bogart helper method to create a JSGI response.
 * Returns a default JSGI response with body containing the specified html, a status of 200,
 * and headers.  The headers included are "content-type" of "text/html" and "content-length" set
 * appropriately based upon the length of the 'html' parameter.
 *
 * @param {String} html  HTML for the body of the response
 * @param {Object} opts  Options to override JSGI response defaults.  For example, passing { status: 404 } would
 *                       cause the resulting JSGI response's status to be 404.
 *
 * @returns JSGI Response
 * @type Object
 */ 
exports.html = function(html, opts) {
  html = html || '';

  return jsgiResponse(html, opts, {
    'content-type': 'text/html',
    'content-length': Buffer.byteLength(html, 'utf-8')
  });
};

/**
 * Bogart helper function to create a JSGI response.
 * Returns a default JSGI response with body containing the specified object represented as JSON, a status of 200,
 * and headers.  The headers included are "content-type" of "application/json" and "content-length" set 
 * appropriately based upon the length of the JSON representation of @paramref(obj)
 *
 *     var resp = bogart.json({ a: 1});
 *     sys.inspect(resp)  
 * 
 * Assuming node-style sys.inspect, evalutes to:
 * { 
 *   status: 200,
 *   headers: { "content-type": "application/json", "content-length": 5 },
 *   body: [ "{a:1}" ]
 * }
 *                               
 *
 * @param {Object} obj  Object to be represented as JSON
 * @param {Object} opts Options to override JSGI response defaults.  For example, passing {status: 404 } would 
 *                      cause the resulting JSGI response's status to be 404.
 */
exports.json = function(obj, opts) {
  opts = opts || {};
  
  var str = JSON.stringify(obj);
  
  return {
    status: opts.status || 200,
    body: [str],
    headers: { "content-type": "application/json", "content-length": Buffer.byteLength(str, 'utf-8') }
  };
};

exports.error = function(msg, opts) {
  opts = opts || {};
  msg = msg || "Server Error";
  
  return {
    status: opts.status || 500,
    body: [msg],
    headers: { "content-type": "text/html", "content-length": Buffer.byteLength(msg, 'utf-8') }
  };
};

/**
 * Bogart helper function to create a JSGI response.
 * Returns a default JSGI response the redirects to the url provided by the 'url' parameter.
 *
 *     var resp = bogart.redirect("http://google.com");
 *     sys.inspect(resp)  
 * 
 * Assuming node-style sys.inspect, evalutes to:
 * { 
 *   status: 302,
 *   headers: { "location": "http://google.com" },
 *   body: []
 * }
 *                               
 *
 * @param {String} url  URL to which the JSGI response will redirect
 * @returns JSGI response for a 302 redirect
 * @type Object
 */
exports.redirect = function(url, opts) {
  var resp = {
    status: 302,
    headers: { "location": url },
    body: []
  };

  if (opts) {
    if (opts.headers) {
      resp.headers = merge(resp.headers, opts.headers);
      delete opts.headers;
    }

    merge(resp, opts);
  }

  return resp;
};

/**
 * Bogart helper function to create a JSGI response.
 * Returns a default JSGI response the redirects to the url provided by the 'url' parameter.
 *
 *     var resp = bogart.permanentRedirect("http://google.com");
 *     sys.inspect(resp)  
 * 
 * Assuming node-style sys.inspect, evalutes to:
 * { 
 *   status: 301,
 *   headers: { "location": "http://google.com" },
 *   body: []
 * }
 *                               
 *
 * @param {String} url  URL to which the JSGI response will redirect
 * @returns JSGI response for a permanent (301) redirect
 * @type Object
 */
exports.permanentRedirect = function(url, opts){
    var resp = {
        status:301,
        headers: {"location": url},
        body: []
    };

    if (opts) {
      if (opts.headers) {
        resp.headers = merge(resp.headers, opts.headers);
        delete opts.headers;
      }

      merge(resp, opts);
    }

    return resp;
};

/**
 * Bogart helper function to create a JSGI response.
 * Returns a default JSGI response with a status of 304 (not modified).
 *
 *     var resp = bogart.notModified();
 * 
 * JSGI Response:
 *
 *     { 
 *       status: 304,
 *       body: []
 *     }
 *                               
 * @returns JSGI response for a not modified response (304).
 * @type Object
 */
exports.notModified = function(opts){
  return merge({
      status: 304,
      body:[]
  }, opts);  
};

var ResponseBuilder = exports.ResponseBuilder = function(viewEngine) {

  var Stream = require('stream')
    , responseBuilder = Object.create(Q.defer(), { then: { get: function() { return this.promise.then.bind(this.promise); } } })
    , forEachCallback
    , forEachDeferred = Q.defer()
    , response = { headers: {}, status: 200, body: {} }
    , waiting = []
    , ended = false
    , resolved = false;
  
  var send = function(data) {
    if (typeof data === 'string') {
      forEachCallback(data);
    } else if (typeof data.forEach === 'function') {
      data.forEach(forEachCallback);
    } else {
      forEachCallback(data);
    }
  };

  responseBuilder.writable = true;

  responseBuilder.on = function(event, callback) {
  };

  responseBuilder.emit = function(event) {
  };

  responseBuilder.removeListener = function(event, callback) {
    
  };

  /**
   * Send response data
   *
   * Examples:
   *
   *    res.send('Hello World');
   *    res.send([ 'Hello', 'World' ]);
   *    res.send(new Buffer('Hello World'));
   *
   * @param {String | ForEachable | Buffer} data  Data to send
   * @api public
   */
  responseBuilder.send = responseBuilder.write = function(data) {
    if (!resolved) {
      responseBuilder.resolve(response);
      resolved = true;
    }

    if (typeof forEachCallback === 'function') {
      send(data);
    } else {
      waiting.push(data);
    }
  };

  /**
   * Render a `view` to the response stream.
   *
   * Example:
   *
   *    res.render('index.html', { locals: { title: 'Hello World' } });
   *
   * @param {String} view  The view to render
   * @param {Object} opts  Options for the ViewEngine.
   */
  responseBuilder.render = function(view, opts) {
    if (!viewEngine) {
      throw new Error("No viewEngine specified");
    }

    viewEngine.render(view, opts).then(function(content) {
      responseBuilder.send(content);
      responseBuilder.end();
    })
  };

  /**
   * End the response.
   */
  responseBuilder.end = function() {
    forEachDeferred.resolve();
  };

  responseBuilder.headers = function(headers) {
    response.headers = headers;
  };

  responseBuilder.setHeader = function(k,v) {
    response.headers = response.headers || {};
    response.headers[k] = v;
  }

  responseBuilder.status = function(status) {
    if (status !== undefined) {
      if (isNaN(status)) {
        throw new Error('status must be a number');
      }

      response.status = status;
      return responseBuilder;
    }

    return response.status;
  };

  Object.defineProperty(responseBuilder, 'statusCode', {
    get: function() {
      return response.status;
    },
    set: function(value) {
      responseBuilder.status(value);
    }
  });

  response.body.forEach = function(callback) {
    forEachCallback = callback;

    if (waiting.length > 0) {
      waiting.forEach(send);
      waiting = [];
    }

    return forEachDeferred.promise;
  };

  return responseBuilder;
};

/**
 * Retrieve a ResponseBuilder to build a JSGI response imperatively.
 *
 *     var viewEngine = bogart.viewEngine('mustache');
 *     app.get('/', function(req) {
 *       var resp = bogart.res(viewEngine);
 *       
 *       doSomethingAsync(function(err, str) {
 *         if (err) {
 *           resp.status(500);
 *           resp.send('Error: '+err.reason);
 *         } else {
 *           resp.send(str);
 *         }
 *         resp.end();
 *       });
 *       
 *       return resp;
 *     });
 *
 * @param   {ViewEngine} viewEngine   The ViewEngine to be used by response helpers for rendering views.
 * @returns {ResponseBuilder}  An object with methods to help build a response.
 */
exports.res = function(viewEngine) {
  return new ResponseBuilder(viewEngine);
};

exports.response = function(viewEngine) {
  console.log('bogart.response() called, bogart.res() is now the preferred, wrist-friendly, ' + 
    'version of this method.  bogart.response() will be removed in the future.');

  return bogart.res(viewEngine);
};

function pipeStream(stream, opts) {
  var response = exports.res();

  opts = opts || {};

  if (!stream.readable) {
    throw "Streams passed to pipe must be readable streams."
  }

  if (opts.status) {
    response.status(opts.status);
  }

  if (opts.headers) {
    response.headers(opts.headers);
  }

  stream.on('data', function(data) {
    response.send(data);
  });

  stream.on('end', function() {
    response.end();
  });

  stream.on('error', function(err) {
    response.reject(err);
  });

  return response.promise;  
};

/**
 * Pipe a response to a JSGI stream
 *
 * @param {ReadableStream} stream  A readable stream.
 * @returns {Promise} A promise for a JSGI stream.
 */
exports.pipe = function(stream, opts) {
  var deferred = Q.defer();

  if (typeof stream.forEach === 'function') {
    deferred.resolve(merge({}, opts, { body: stream }));
    return deferred.promise;
  } else {
    return pipeStream(stream, opts);
  }
};

/**
 * Get MIME type for a file extension
 *
 * @param {String} ext  File extension
 * @returns {String}  MIME type of file extension.
 */
exports.mimeType = function(ext) {
  var dotIndex = ext.lastIndexOf('.');
  if (dotIndex > 0) {
    ext = ext.substring(dotIndex);
  }

  return require("./mimetypes").mimeType(ext);
};

/**
 * Creates a JSGI response that streams a file
 *
 * @param {String} filePath  The path to the file to be streamed
 * @param {Object} opts      JSGI options
 *
 * @returns {Promise} A promise for a JSGI response
 */
exports.file = function(filePath, opts) {
  opts = opts || {};
  opts.headers = opts.headers || {};
  opts.headers['Content-Type'] = opts.headers['Content-Type'] || exports.mimeType(path.extname(filePath));

  return exports.pipe(fs.createReadStream(filePath), opts);
};

exports.proxy = function(url) {
  return exports.pipe(require('request')(url));
};

/**
 * Helper function to determine the main directory of the application
 * @returns {String} Directory of the script that was executed
 */
exports.maindir = function() {
  if (typeof require.main === 'undefined') {
    return __dirname;
  }
  return path.dirname(require.main.filename).replace("file://","");
};

/**
 * An empty function. Useful for authors of APIs with optional callbacks.
 */
exports.noop = function(){};

/**
 * Pipes data from source to dest.
 *
 * @param {ForEachable | ReadableStream} src   Source of data
 * @param {WriteableStream}              dest  Write data from src to dest
 *
 * @returns {Promise}  A promise that will be resolved when the pumping is completed
 */
exports.pump = require('./stream').pump;

view.setting('template directory', path.join(exports.maindir(), 'views'));
